### PR TITLE
lppublish, cmd/charmload: ingest exact revisions from legacy charm store

### PR DESCRIPTION
**Pull request by rogpeppe**
This behaviour is controlled by the new "-legacy" flag to charmload, which should
be used just once, when initially uploading charms.

In passing, also adjusted log messages somewhat so the logging at INFO
level is a little quieter and more informative.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/331/

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/332/

File: lppublish/legacy.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	path := "meta/revision-info?" + strings.Join(ids, "&")
+
+	var resp map[string]params.RevisionInfoResponse
+	if err := cl.charmStoreGet(path, &resp); err != nil {
+		if errgo.Cause(err) != params.ErrNotFound {
+			return nil, errgo.Notef(err, "cannot retrieve revisions")
+		}
+		return nil, nil
+	}
```
Comment (urosj) [inline lppublish/legacy.go:68]: So, it is not an error if it returns "not found" due to the "design" of the legacy charm store?

File: lppublish/legacy.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	hasher := sha512.New384()
+	size, err := io.Copy(hasher, f)
+	if err != nil {
+		return errgo.Notef(err, "cannot read charm archive: %v", err)
+	}
+
+	// Upload content to all URLs.
+	hash := fmt.Sprintf("%x", hasher.Sum(nil))
+	for _, url := range need {
+		_, err := f.Seek(0, 0)
```
Comment (urosj) [inline lppublish/legacy.go:120]: Smart reuse!

**Comment by jrwren** [comment]
It looks like it uses the branch tips from launchpad as the list of charms to get from the legacy store. Do we know if there is anything in the store which was put there from a now deleted launchpad branch tip?

:+1:

File: lppublish/lpad.go [hunk 3]
```diff
@@ -509,15 +528,14 @@ func (cl *charmLoader) archiveDir(archiver archiverTo, tempDir string) (archiveF
 	return f, hash, fileInfo.Size(), nil
 }
 
-func (cl *charmLoader) postArchive(r io.Reader, id *charm.Reference, size int64, hash string) (*charm.Reference, error) {
+func (cl *charmLoader) uploadArchive(method string, r io.Reader, id *charm.Reference, size int64, hash string) (*charm.Reference, error) {
```
Comment (mhilton) [inline lppublish/lpad.go:533]: Do we expect to use different methods at some point?
Edit: ignore that I missed it earlier

**Comment by mhilton** [comment]
:+1: LGTM

**Comment by urosj** [comment]
:+1: Awesome, @rogpeppe , @frankban .

We do need to start with a clean DB to get all revisions and that is a good requirement!

File: lppublish/legacy.go (hunk 1 above)
Comment (rogpeppe) [inline lppublish/legacy.go:68]: Good question.

Actually, it shouldn't return a not-found error at all for the bulk meta endpoint.
The returned map should just omit entries that were not found.
I'll remove this logic in a subsequent branch.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/charmstore-merge/112

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/340/

**Pull request merged**